accepted/0000-validate-signatures.md

```diff
@@ -0,0 +1,510 @@
+# Improve npm signature verification
+
+## Summary
+
+Add a new top-level CLI command to verify registry signatures.
+
+## Motivation
+
+Signatures are only useful if people verify them. Signature verification is
+currently a [manual and complex process](https://docs.npmjs.com/verifying-the-pgp-signature-for-a-package-from-the-npm-public-registry),
+involving the Keybase CLI to fetch the npm public key.
+
+Our long-term goal for supply chain security is that all software is signed and
+verified in a transparent and user-friendly way.
+
+Signing and verifying published packages protects users against a malicious
+mirror or proxy intercepting and tamptering with the package response (MITM
+attack). Mirrors and proxyies are common in the npm ecosystem, significantly
+increasing the possible attack surface.
+
+## Detailed Explanation
+
+Introduce a new CLI command `verify-signatures` that verifies the npm signature
+in a packages packument, using npm's public key, fetched from npmjs.org. It
+works on the current install (`node_modules`), and checks all direct and
+transitive dependencies.
+
+The aim is for this command to plug into users build workflows after `npm ci/npm
+install` and block builds with invalid signatures.
+
+This command is standalone but we could fold this behaviour into `npm install`
+or `audit` in the future once we're confident validation is performant and
+provides a good user experience with little to no noise.
+
+### Detailed CLI examples
+
+For the next few examples, assume an install such as:
+
+```sh
+$ npm ls
+project@1.0.0 $HOME/work/project
+├── foo@0.4.0
+├─┬ lorem@0.4.0
+│ └── ipsum@2.0.0 invalid signatures
+├─┬ abbrev@3.0.9
+│ └── bar@2.88.0 invalid signatures
+└── once@1.4.0 invalid signatures
+```
+
+#### Verifies all dependencies in the current install, e.g:
+
+```
+$ npm verify-signatures
+project@1.0.0 $HOME/work/project
+
+1 package has a missing signature (direct):
+├── MISSING SIGNATURE foo@0.4.0
+
+NOTE: Signatures are missing for all packages hosted outside registry.npmjs.org
+
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+!!  WARNING: SOME PACKAGES DOWNLOADED FROM NPMJS.ORG HAVE INVALID SIGNATURES  !!
+!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+3 packages have invalid signatures (1 direct, 2 transitive):
+├── INVALID SIGNATURE once@1.4.0
+├─┬ lorem@0.4.0
+│ ├── INVALID SIGNATURE ipsum@2.0.0
+├─┬ abbrev@3.0.9
+│ ├── INVALID SIGNATURE bar@2.88.0
+
+Someone might have tampered with the package since it was hosted
```

Comment:
I think we should tweak some of the wording here a bit, as `since it was hosted` is a bit ambiguous to me.

Side question: do we do i18n for the CLI, or is English hard-coded?